### PR TITLE
Feature: GameDig Widget

### PR DESCRIPTION
**Pull request by fredodiable**
## Proposed change

Adding Counter-Strike: Global Offensive (CSGO) widget
It use [gamedig](https://www.npmjs.com/package/gamedig) library to get game's server informations
Closes [#1696](https://github.com/benphelps/homepage/discussions/1696)

CSGO widget

## Type of change

- [x] New service widget
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (please explain)

## Checklist:

- [x] If adding a service widget or a change that requires it, I have added a corresponding PR to the [documentation](https://github.com/benphelps/homepage-docs) here: [#128](https://github.com/benphelps/homepage-docs/pull/128)
- [x] If adding a new widget I have reviewed the [guidelines](https://gethomepage.dev/en/more/development/#service-widget-guidelines).
- [x] If applicable, I have checked that all tests pass with e.g. `pnpm lint`.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.

### CODE REVIEW

**Comment by benphelps** [comment]
I think this dependency has enough possible utility to be included.  However, since the gamedig response object is normalized to contain a predictable set of keys, I think this has huge opportunity to be generalized into any game server that it supports.

Could you change this so that the game server type is also passed as an argument, and add the Blocks for `maxplayers` and `name`.

**Comment by benphelps** [comment]
You'll need to rename the widget from CSGO to just GameDig, and update/add the i18n keys for `maxplayers` and `name`.

**Comment by shamoon** [comment]
That’s an awesome idea

**Comment by fredodiable** [comment]
I converted csgo widget to gamedig
Added game server type as widget parameter
Added extra fields to let the user cutomize the widget

**Comment by shamoon** [comment]
Awesome. One snag is there still needs to be a 4 block max https://gethomepage.dev/en/more/development/ hmm

**Comment by benphelps** [comment]
> Awesome. One snag is there still needs to be a 4 block max https://gethomepage.dev/en/more/development/#service-widget-guidelines hmm

Yeah, I think the best option is to provide 3 default fields in the widget docs, for copy-pasting, then users can choose the ones they want to see afterwords.  Otherwise we'll for sure get requests to add them anyways. :|

**Comment by fredodiable** [comment]
That's what I was thinking too
Doc already as 3 fields, the most useful IMO
See first row layout in my screenshot
Is this OK for you ?

**Comment by benphelps** [comment]
Let's see if @shamoon has any better ideas, if not, we'll get this merged in.

**Comment by shamoon** [comment]
It does complicate the logic a bit, of course, but for example the nextcloud widget supports > 4 field choices but display is limited to 4. That would be my approach, so users can choose which they like but maintain the legibility / layout of the widget (the purpose of the 4 limit)

**Comment by fredodiable** [comment]
> It does complicate the logic a bit, of course, but for example the nextcloud widget supports > 4 field choices but display is limited to 4. That would be my approach, so users can choose which they like but maintain the legibility / layout of the widget (the purpose of the 4 limit)

I limited widgets displayed fields globally to 4
I added 3 default fields for gamedig widget

**Comment by shamoon** [comment]
Thanks for being so flexible! Is this something I can easily use a free account somewhere to test? Alternatively could you post an example of the api output?

File: src/utils/config/service-helpers.js
Comment (shamoon) [inline src/utils/config/service-helpers.js:1]: Oh actually we had something like this before and it caused a lot of issues, annoying as it is, for now this logic should be in the widget itself

**Comment by fredodiable** [comment]
> Thanks for being so flexible! Is this something I can easily use a free account somewhere to test? Alternatively could you post an example of the api output?

You can use any server from here https://www.battlemetrics.com/servers
I tested these games
- Counter-Strike: Global Offensive
- Team Fortress 2
- Minecraft
- Rust
- Arma 3

For some servers you will need to use the query port

File: src/utils/config/service-helpers.js
Comment (fredodiable) [reply src/utils/config/service-helpers.js:1]: > Oh actually we had something like this before and it caused a lot of issues, annoying as it is, for now this logic should be in the widget itself

Done, moved fields limit to gamedig widget

**Review by shamoon — APPROVED** [review]
Awesome

Screenshot 2023-07-26 at 10 05 18 PM
Screenshot 2023-07-26 at 10 06 18 PM

**Comment by mgrimace** [comment]
Can gamedig pull `status` with response `online` or `offline` for servers such as Minecraft? If so, could that be added as a possible field? Thanks. FYI this is what works for me for Minecraft bedrock edition: 

```
    - Minecraft Test Widget:
        icon: minecraft.png
        description: Ubuntu (Oracle)
        widget:
            type: gamedig
            serverType: minecraftbe
            url: udp://[your game server IP]:19132
            fields: ["name", "players", "ping"]
```
Screenshot 2023-08-05 at 7 18 51 PM

**Comment by fredodiable** [comment]
> Can gamedig pull `status` with response `online` or `offline` for servers such as Minecraft? If so, could that be added as a possible field? Thanks. FYI this is what works for me for Minecraft bedrock edition:
> 
> ```
>     - Minecraft Test Widget:
>         icon: minecraft.png
>         description: Ubuntu (Oracle)
>         widget:
>             type: gamedig
>             serverType: minecraftbe
>             url: udp://[your game server IP]:19132
>             fields: ["name", "players", "ping"]
> ```
> 
> Screenshot 2023-08-05 at 7 18 51 PM

Sure, let me some days, I'm not at home right now

**Comment by fredodiable** [comment]
Change done, waiting for pull request [#1789](https://github.com/benphelps/homepage/pull/1789)

**Pull request merged**